asmcomp/power/selection.ml

```diff
@@ -43,13 +43,28 @@ let rec select_addr = function
   | exp ->
       (Alinear exp, 0, Debuginfo.none)
 
+let is_immediate n = n <= 0x7FFF && n >= -0x8000
+let is_immediate_logical n = n <= 0xFFFF && n >= 0
+
 (* Instruction selection *)
 
 class selector = object (self)
 
 inherit Selectgen.selector_generic as super
 
-method is_immediate n = (n <= 32767) && (n >= -32768)
+method is_immediate_test cmp n =
+  match cmp with
+  | Isigned _ -> is_immediate n
+  | Iunsigned _ -> is_immediate_logical n
+
+method! is_immediate op n =
+  match op with
+  | Iadd | Imul -> is_immediate n
+  | Isub -> is_immediate (-n)  (* turned into add opposite *)
+  | Iand | Ior | Ixor -> is_immediate_logical n
+  | Icomp c -> self#is_immediate_test c n
+  | Icheckbound _ -> is_immediate_logical n   (* unsigned comparison *)
```

Comment:
I think is `is_immediate` was better than `is_immediate_logical`, although neither is quite right.

The `tdlle` / `twlle` instruction that implements checkbound takes a 16-bit immediate argument. The comparison is unsigned, but the immediate is sign-extended (!). So the representable immediates are (on 64-bit):
```
0 <= n <= 0x7FFF
0xFFFF_FFFF_FFFF_8000 <= n <= 0xFFFF_FFFF_FFFF_FFFF
```
Since there can't be arrays long enough to hit the second case, `is_immediate` probably does the right thing, but `0 <= n <= 0x7FFF` seems like the right test to put here.

Reply:
Right, `is_immediate_logical` causes errors in testing, I thought I already reverted to `is_immediate`.  But your suggestion `0 <= n <= 0x7FFF` is better, I'll implement this ASAP.